charging_station.py

```diff
@@ -0,0 +1,155 @@
+        # p_max and c_max depend on whether the car is here or not.
+        for speed in ["slow","fast"] :
+            for i in range(2):
+                if abs(load_battery[speed][i]) >= p_max[speed][i]:
+                    load_battery[speed][i] = p_max[speed][i]*np.sign(load_battery[speed][i])
+            # Can't put more power than p_max
+            new_stock = { "slow" : [0,0], "fast" : [0,0] }
+
+            new_stock["slow"][0] = self.battery_stock["slow"][time][0] + (self.efficiency*max(0,load_battery["slow"][0])+min(0,load_battery["slow"][0])/self.efficiency)*self.dt
+            new_stock["slow"][1] = self.battery_stock["slow"][time][1] + (self.efficiency*max(0,load_battery["slow"][1])+min(0,load_battery["slow"][1])/self.efficiency)*self.dt
+            # We update the new stock of each batteries "slow"
+
+            new_stock["fast"][0]=self.battery_stock["fast"][time][0] + (self.efficiency*max(0,load_battery["fast"][0])+min(0,load_battery["fast"][0])/self.efficiency)*self.dt
+            new_stock["fast"][1]=self.battery_stock["fast"][time][1] + (self.efficiency*max(0,load_battery["fast"][1])+min(0,load_battery["fast"][1])/self.efficiency)*self.dt
+            # We update the new stock of each batteries "fast"
+
+
+        for speed in ["slow","fast"] :
+            for i in range(2):
+                if new_stock[speed][i] < 0:
+                    load_battery[speed][i] = -(self.battery_stock[speed][time][i])/(self.efficiency*self.dt)
+                    new_stock[speed][i] = 0
+            # We can't discharge the batterie under 0
+
+                elif new_stock[speed][i] > c_max[speed][i]:
+                    load_battery[speed][i] = (c_max[speed][i] - self.battery_stock[speed][time][i] ) / (self.efficiency*self.dt)
+                    new_stock[speed][i] = c_max[speed][i]
+            # We can't charge the batteries over their maximum capacities
+
+        for speed in ["slow","fast"] :
+            for i in range(2):
+                if abs(load_battery[speed][i]) >= p_max[speed][i] :
+                    load_battery[speed][i] = p_max[speed][i]*np.sign(load_battery[speed][i])
+
+        for speed in ["slow","fast"] :
+            for i in range(2):
+                self.battery_stock[speed][time+1][i]=new_stock[speed][i]
+            # Update of batteries stocks
+                if time == self.arrival[speed][i]-1:
+                    self.battery_stock[speed][time+1][i] = self.battery_stock[speed][int(self.depart[speed][i])][i]-4
+                    # When the cars comes back it has lost 4 kWh in the battery
+                self.load_battery_periode[speed][time][i] = load_battery[speed][i]
+
+        return load_battery # We return load_battery clear of the player's potential mistakes
+
+
+
+
+    def nb_cars(self,time):
+        s = 0
+        for i in range(self.nb_slow_max):
+            if (self.depart["slow"][i]<time) and (self.arrival["slow"][i]>time):
```

Comment:
This condition seems strange to me: EV are connected before departure or after arrival in our setting

Reply:
oui désolé j'ai inversé le sens des inégalités donc j'ai modifié here en conséquence mais j'ai oublié de modifier s et f dont on a pas vraiment besoins en réalité à part peut etre pour la visualisation ? donc je les gardes ou pas (nb_slow et nb_fast en temps réel)

Reply:
Cela serait bien de faire cette correction pour avoir les valeurs correctes de nb_slow et nb_fast (cela ne coûte pas grand chose !). Elles pourront peut-être servir pour la visualisation effectivement